ext/EnzymeExt.jl

```diff
@@ -1,151 +1,382 @@
+    args...,
+) where {ModifiedBetween,FT,TapeType}
+    # A2 = Const{Nothing} -- since f->Nothing
+    forward, _ = EnzymeCore.autodiff_deferred_thunk(
+        ReverseSplitModified(ReverseSplitWithPrimal, Val(ModifiedBetween)),
+        TapeType,
+        Const{Core.Typeof(f)},
+        Const{Nothing},
+        Const{Core.Typeof(ctx)},
+        map(Core.Typeof, args)...,
+    )
 
-        aug_kernel = similar(kernel, aug_fwd_cpu)
+    # On the GPU: F is a per thread function
+    # On the GPU: subtape::Vector
+    I = __index_Global_Linear(ctx)
+    subtape[I] = forward(Const(f), Const(ctx), args...)[1]
+    return nothing
+end
 
-        aug_kernel(f, ModifiedBetween, subtape, args2...; ndrange, workgroupsize)
+function gpu_rev(
+    ctx,
+    f::FT,
+    ::Val{ModifiedBetween},
+    subtape,
+    ::Val{TapeType},
+    args...,
+) where {ModifiedBetween,FT,TapeType}
+    # XXX: TapeType and A2 as args to autodiff_deferred_thunk
+    _, reverse = EnzymeCore.autodiff_deferred_thunk(
+        ReverseSplitModified(ReverseSplitWithPrimal, Val(ModifiedBetween)),
+        TapeType,
+        Const{Core.Typeof(f)},
+        Const{Nothing},
+        Const{Core.Typeof(ctx)},
+        map(Core.Typeof, args)...,
+    )
+    I = __index_Global_Linear(ctx)
+    tp = subtape[I]
+    reverse(Const(f), Const(ctx), args..., tp)
+    return nothing
+end
 
-        # TODO the fact that ctxTy is type unstable means this is all type unstable.
-        # Since custom rules require a fixed return type, explicitly cast to Any, rather
-        # than returning a AugmentedReturn{Nothing, Nothing, T} where T.
+function EnzymeRules.augmented_primal(
+    config::Config,
+    func::Const{<:Kernel},
+    ::Type{Const{Nothing}},
+    args::Vararg{Any,N};
+    ndrange = nothing,
+    workgroupsize = nothing,
+) where {N}
+    kernel = func.val
+    f = kernel.f
 
-        res =  AugmentedReturn{Nothing, Nothing, Tuple{Array, typeof(arg_refs)}}(nothing, nothing, (subtape, arg_refs))
-        return res
-    end
+    ndrange, workgroupsize, iterspace, dynamic =
+        launch_config(kernel, ndrange, workgroupsize)
+    ctx = _enzyme_mkcontext(kernel, ndrange, iterspace, dynamic)
+    ctxTy = Core.Typeof(ctx) # CompilerMetadata{ndrange(kernel), Core.Typeof(dynamic)}
+    # TODO autodiff_deferred on the func.val
+    ModifiedBetween = Val((overwritten(config)[1], false, overwritten(config)[2:end]...))
 
-    function EnzymeRules.reverse(config::Config, func::Const{<:Kernel{CPU}}, ::Type{<:EnzymeCore.Annotation}, tape, args::Vararg{Any, N}; ndrange=nothing, workgroupsize=nothing) where N
-        subtape, arg_refs = tape
+    FT = Const{Core.Typeof(f)}
 
-        args2 = ntuple(Val(N)) do i
-            Base.@_inline_meta
-            if args[i] isa Active
-                EnzymeCore.MixedDuplicated(args[i].val, arg_refs[i])
+    arg_refs = ntuple(Val(N)) do i
+        Base.@_inline_meta
+        if args[i] isa Active
+            if func isa Const{<:Kernel{<:GPU}}
```

Comment:
I'm confused by this, why is this check needed for func?

Reply:
Because `Active` arguments will throw a cryptic error that doesn't make it clear that `Active` is the issue.

Reply:
Can you make it func.Val isa Kernel GPU then?

Reply:
Done